### PR TITLE
Javascript: Add new queries for Javascript actions

### PR DESCRIPTION
**Pull request by R3x**
Added 3 Queries, that track flows from - 

1. Action Input to Sinks
2. Action Github Context to Sinks
3. Action Environment to Sinks

### CODE REVIEW

**Comment by R3x** [comment]
@kevinbackhouse @JarLob - I have added the sinks that are of higher priority here. During the analysis - we also marked some lower-priority sinks (things such as `fs.write` etc). Which basically provides primitives such as an arbitrary file write - but not as severe as the ones here. Do you prefer that I add them as well?

**Comment by kevinbackhouse** [comment]
> @kevinbackhouse @JarLob - I have added the sinks that are of higher priority here. During the analysis - we also marked some lower-priority sinks (things such as `fs.write` etc). Which basically provides primitives such as an arbitrary file write - but not as severe as the ones here. Do you prefer that I add them as well?

I think it's better to start with just the high-priority ones, so that you have a lower false-positive rate.

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+DataFlow::Node discussionSources() {
+  result = payloadObject("discussion", "title") or
+  result = payloadObject("discussion", "body")
+}
+
+DataFlow::Node commentSources() {
+  result = payloadObject("comment", "body") or
+  result = payloadObject("review", "body") or
+  result = payloadObject("review_comment", "body")
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:71]:
```suggestion
  result = payloadObject("comment", "body")
```
I think the correct payload is `github.event.comment.body`

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  result =
+    DataFlow::moduleImport("@actions/github")
+        .getAPropertyRead("context")
+        .getAPropertyRead("payload")
+        .getAPropertyRead(event)
+}
+
+DataFlow::Node pullrequestSources() {
+  result = payloadObject("pull_request", "head").getAPropertyRead("ref") or
+  result = payloadObject("pull_request", "head").getAPropertyRead("label") or
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:53]:
```suggestion
  result = payloadObject("pull_request", "head").getAPropertyRead("label") or
  result = payloadObject("pull_request", "head").getAPropertyRead("repo").getAPropertyRead("default_branch") or
  result = payloadObject("pull_request", "head").getAPropertyRead("repo").getAPropertyRead("description") or
  result = payloadObject("pull_request", "head").getAPropertyRead("repo").getAPropertyRead("homepage") or
```

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+DataFlow::Node headCommitSources() {
+  result = payloadObject("head_commit", "message") or
+  result = payloadObject("head_commit", "author").getAPropertyRead("name") or
+  result = payloadObject("head_commit", "author").getAPropertyRead("email") or
+  result = payloadEvent("commits").getAPropertyRead().getAPropertyRead("message") or
+  result =
+    payloadEvent("commits").getAPropertyRead().getAPropertyRead("author").getAPropertyRead("name") or
+  result =
+    payloadEvent("commits").getAPropertyRead().getAPropertyRead("author").getAPropertyRead("email")
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:99]:
```suggestion
  result =
    payloadEvent("commits").getAPropertyRead().getAPropertyRead("author").getAPropertyRead("name") or
  result =
    payloadEvent("commits").getAPropertyRead().getAPropertyRead("author").getAPropertyRead("email") or
   result = payloadEvent("commits").getAPropertyRead().getAPropertyRead("committer").getAPropertyRead("name") or
  result = payloadEvent("commits").getAPropertyRead().getAPropertyRead("committer").getAPropertyRead("email")
```

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        .getAPropertyRead("pull_requests")
+        .getAPropertyRead()
+        .getAPropertyRead("head")
+        .getAPropertyRead("ref")
+}
+
+DataFlow::Node headCommitSources() {
+  result = payloadObject("head_commit", "message") or
+  result = payloadObject("head_commit", "author").getAPropertyRead("name") or
+  result = payloadObject("head_commit", "author").getAPropertyRead("email") or
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:94]:
```suggestion
  result = payloadObject("head_commit", "author").getAPropertyRead("name") or
  result = payloadObject("head_commit", "author").getAPropertyRead("email") or
  result = payloadObject("head_commit", "committer").getAPropertyRead("name") or
  result = payloadObject("head_commit", "committer").getAPropertyRead("email") or
```

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+DataFlow::Node workflowRunSources() {
+  result = payloadObject("workflow_run", "head_branch") or
+  result = payloadObject("workflow_run", "head_commit").getAPropertyRead("message") or
+  result =
+    payloadObject("workflow_run", "head_commit").getAPropertyRead("author").getAPropertyRead("name") or
+  result =
+    payloadObject("workflow_run", "head_commit")
+        .getAPropertyRead("author")
+        .getAPropertyRead("email") or
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:82]:
```suggestion
  result =
    payloadObject("workflow_run", "head_commit")
        .getAPropertyRead("author")
        .getAPropertyRead("email") or
    result =
    payloadObject("workflow_run", "head_commit")
        .getAPropertyRead("committer")
        .getAPropertyRead("email") or
```

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  result = payloadObject("comment", "body") or
+  result = payloadObject("review", "body") or
+  result = payloadObject("review_comment", "body")
+}
+
+DataFlow::Node workflowRunSources() {
+  result = payloadObject("workflow_run", "head_branch") or
+  result = payloadObject("workflow_run", "head_commit").getAPropertyRead("message") or
+  result =
+    payloadObject("workflow_run", "head_commit").getAPropertyRead("author").getAPropertyRead("name") or
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:78]:
```suggestion
  result =
    payloadObject("workflow_run", "head_commit").getAPropertyRead("author").getAPropertyRead("name") or
    result =
    payloadObject("workflow_run", "head_commit").getAPropertyRead("committer").getAPropertyRead("name") or
```

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+DataFlow::Node commentSources() {
+  result = payloadObject("comment", "body") or
+  result = payloadObject("review", "body") or
+  result = payloadObject("review_comment", "body")
+}
+
+DataFlow::Node workflowRunSources() {
+  result = payloadObject("workflow_run", "head_branch") or
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:75]:
```suggestion
  result = payloadObject("workflow_run", "head_branch") or
  result = payloadObject("workflow_run", "display_title") or
  result = payloadObject("workflow_run", "head_repository").getAPropertyRead("description") or
```

**Review by JarLob — CHANGES_REQUESTED** [review]
Reformatting of the added lines may be needed.

File: javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+DataFlow::Node discussionSources() {
+  result = payloadObject("discussion", "title") or
+  result = payloadObject("discussion", "body")
+}
+
+DataFlow::Node commentSources() {
+  result = payloadObject("comment", "body") or
+  result = payloadObject("review", "body") or
+  result = payloadObject("comment", "body")
```
Comment (JarLob) [inline javascript/ql/src/experimental/Security/CWE-094/ActionsContextToSink.ql:74]:
```suggestion
  result = payloadObject("review", "body")
```
Sorry, my previous suggestion actually introduced the duplicate.

**Comment by asgerf** [comment]
Hi @R3x and @JarLob 👋 

First of all, sorry for the long radio silence.

These models seem very valuable, and I'd like to get them into the mainline outside experimental. However, I foresee some issues with code duplication, since this PR adds new variants of existing queries containing sinks we already have models for. It's difficult for us to maintain code with such duplication. Rather than merging to experimental and then do a round of clean-up afterwards, I've opened a PR adding the new sources as well as `@actions/exec` to mainline.

I haven't ported the "Actions Environment to Sink" query as I'm still not quite sure what the threat model is, but I'm thinking it could be part of `js/shell-command-injection-from-environment`

**Comment by R3x** [comment]
Closing this. since we have #12978

**Pull request closed without merging**